src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,399 @@
+            if ((info = anExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(pseudoRandomKey, info, length),
+                                     length);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "an HKDF Expand could not be initialized with the given "
+                    + "key material").initCause(ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationParameterSpec instanceof HKDFParameterSpec.ExtractThenExpand) {
+            HKDFParameterSpec.ExtractThenExpand anExtractThenExpand =
+                (HKDFParameterSpec.ExtractThenExpand) derivationParameterSpec;
+            ikms = anExtractThenExpand.ikms();
+            salts = anExtractThenExpand.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // set this value in the "if"
+            if ((info = anExtractThenExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExtractThenExpand.length();
+            if (length > (hmacLen * 255)) {
+                throw new InvalidAlgorithmParameterException(
+                    "Requested length exceeds maximum allowed key stream "
+                    + "length");
+            }
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(inputKeyMaterial, (salt
+                                                                      == null) ? null : salt.getEncoded());
+                pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
```

Comment:
Use `hmacAlgName` instead of `RAW` as the key algorithm? Or is it well-established fact that most if not all Hmac impls doesn't care or check the key algorithm?

Reply:
Valerie is correct. I might have confused algorithm name and format name.

Reply:
Yes, resolved.